Add az-security-definitions rule

mikekistler opened the pull request:
This PR adds the az-security-definitions rule which checks that security schemes in `securityDefinitions` contain the information needed by autorest to produce client libraries with the correct authentication features.

This is a new requirement for data plane codegen and currently most Azure services are not in compliance, particularly in regard to the presence of scopes and pattern of scope names.

MushMal commented on functions/security-definitions.js:
```diff
@@ -0,0 +1,67 @@
+    return [];
+  }
+
+  if (!doc.securityDefinitions || (typeof doc.securityDefinitions === 'object' && Object.keys(doc.securityDefinitions).length === 0)) {
```
Is there a need for a check if doc.securityDefinitions is not an object type?

MushMal commented on functions/security-definitions.js:
```diff
@@ -0,0 +1,67 @@
+        }
+      } else {
+        errors.push({
+          message: 'Security scheme must be type: oauth2 or type: apiKey.',
```
Nit: perhaps include the name of the scheme in the message for quick matching for the output

mikekistler replied:
The scheme name will be part of the path, which is shown in the linter output.  E.g. "Basic" in the output here:
```
  17:15      warning  az-security-definitions             Security scheme must be type: oauth2 or type: apiKey.        securityDefinitions.Basic.type
```

MushMal commented on functions/security-definitions.js:
```diff
@@ -0,0 +1,67 @@
+
+  Object.keys(schemes).forEach((schemeKey) => {
+    const scheme = schemes[schemeKey];
+    if (typeof scheme === 'object') {
```
I assume there is some validator for the schema overall at a higher level? In case if scheme is not an object for example.

mikekistler replied:
Yes, but I'll add a comment here to avoid confusion.

MushMal commented on test/security-definitions.test.js:
```diff
@@ -0,0 +1,160 @@
+  return linter.run(oasDoc).then((results) => {
+    expect(results.length).toBe(0);
+  });
+});
```
Would be nice to have a composite test with multiple oauth2 and apikey securityDefinitions like the last test with multiple valid and invalid permutations (with scopes, in header, type) in one focument.

MushMal reviewed (commented):
Looks good. Some questions mostly for my understanding of the flow.

My larger concern is whether we should allow for an implicit default value if most services do not specify this in the first place?

mikekistler commented:
@MushMal I pushed a new commit that think addresses your comments.

Regarding a default security definition, I think we should avoid this.  It's better to be explicit and OpenAPI does not even have a means to specify a default.
